design/resource-dependency.md

```diff
@@ -0,0 +1,348 @@
+# Resource Dependency in OAM
+
+- Owner: Hongchao Deng (@hongchaodeng)
+- Date: 05/17/2020
+- Status: Draft
+
+## Background
+
+Creating one application service or resource often relies on other services or resources to be ready.
+This is a well-known pattern called dependency.
+
+In OAM, we talk about dependency in two different contexts:
+
+1. **Inter ApplicationConfiguration**:
+  There is dependency workflow between multiple ApplicationConfigurations. This usually can be achieved
+  via external workflow engine such as Tekton/Argo Workflow.
+2. **Intra ApplicationConfiguration**: 
+  There is dependency between Components and Traits within the same ApplicationConfiguration.
+  Since they are applied at the same time, external workflow engine couldn't do anything to control their ordering.
+  It is the OAM runtime that needs to satisfy this requirement.
+
+In this proposal, we are handling the second case, *Intra ApplicationConfiguration* dependency.
+
+## Goals
+
+This proposal includes the following goals:
+
+1. **Describe ordering for k8s-style resources**.
+  The ordering between resources means creating one resource after another.
+  For example, resource A could not be created until resource B has been created and provided outputs.
+
+2. **Describe field data passing between resources**.
+  It is still not enough to solve only ordering problem. We also need to handle field data passing.
+  For example, resource A's field "spec.connSecret" relies on resource B's field "status.connSecret". 
+  This means the field data of one resource is passed from another field of other resource and needs to wait the field data to be ready.
+  In fact, ordering in the first goal can be modeled as data dependency as well except that the receiver doesn't use the data.
+
+3. **OAM native experience**.
+  OAM has application concepts like Components, Traits, etc.
+  The solution should provide OAM native experience and written as native fields.
+  But OAM resources are k8s resources too.
+  The core dependency engine should be generic enough to extend to more OAM resources.
+
+In the following, we will go through the use cases first and then propose a solution to satisfy them.
+
+
+## Use Cases
+
+### 1. Web app and database
+
+An applicationConfiguration consists of two components: web and db (database).
+Web component should not be created until db component is created,
+including public endpoint and connection secret to be made ready.
+
```

Comment:
We should also consider traditional REST APIs (need the URIs of my other services) as a keystone scenario for any depenedency use case. It's exceedingly common (we already have a sample) and it seems like we could avoid the incredible level of coupling and boilerplate it requires today: https://github.com/oam-dev/samples/blob/master/2.ServiceTracker_App/ApplicationConfiguration/tracker-app-config.yaml#L34

The proposal here seems like it could already help with this, but I want to make sure we're considering it as a headline scenario.

Reply:
That's the purpose of this proposal.